File: c++/functions/passing_arguments.cpp
#include <iostream>

void function_receiving_argument_by_value(int argument)
{
  std::cout << "Function receiving argument by value\n"
    << "-- begin:\n"
    << "before: argument = " << argument << std::endl
    << "argument = argument * 2\n";

  argument *= 2;

  std::cout << "after: arguemnt = " << argument << std::endl
    << "-- end.\n";
}

void function_receiving_argument_by_pointer(int *argument)
{
  std::cout << "Function receiving argument by pointer\n"
    << "-- begin:\n"
    << "before: argument = " << *argument << std::endl
    << "argument = argument * 2\n";

  *argument *= 2;

  std::cout << "after: arguemnt = " << *argument << std::endl
    << "-- end.\n";
}

void function_receiving_argument_by_reference(int &argument)
{
  std::cout << "Function receiving argument by reference\n"
    << "-- begin:\n"
    << "before: argument = " << argument << std::endl
    << "argument = argument * 2\n";

  argument *= 2;

  std::cout << "after: arguemnt = " << argument << std::endl
    << "-- end.\n";
}

int main()
{
  int value = 5;
  int *pointer = &value;
  int &reference = value;

  std::cout << "BEFORE: value = " << value << std::endl;
  function_receiving_argument_by_value(value);
  std::cout << "AFTER: value = " << value << std::endl
  << std::endl;

  std::cout << "BEFORE: value = " << value << std::endl;
  function_receiving_argument_by_pointer(pointer);
  std::cout << "AFTER: value = " << value << std::endl
  << std::endl;

  std::cout << "BEFORE: value = " << value << std::endl;
  function_receiving_argument_by_reference(reference);
  std::cout << "AFTER: value = " << value << std::endl
  << std::endl;
}

File: python/functions/passing_arguments.py
#!/usr/bin/python3

import typing

def function_receiving_value(argument: int) -> None:
    print("Function receiving value\n"
        + "-- begin:\n"
        + f"before: argument = {argument}\n"
        + "argument = argument * 2")

    argument *= 2

    print(f"after: argument = {argument}\n"
        + "-- end.")

def function_receiving_list(argument: list) -> None:
    print("Function receiving list\n"
        + "-- begin:\n"
        + f"before: argument[0] = {argument[0]}\n"
        + "argument[0] *= 2")

    argument[0] *= 2

    print(f"after: argument[0] = {argument[0]}\n"
        + "-- end.")

def function_receiving_tuple(argument: tuple) -> None:
    print("Function receiving tuple\n"
        + "-- begin:\n"
        + f"before: argument[0] = {argument[0]}\n"
        + "argument[0] *= 2")

    argument[0] *= 2 #error

    print(f"after: argument[0] = {argument[0]}\n"
        + "-- end.")

def function_receiving_object(argument: object) -> None:
    print("Function receiving value by value\n"
        + "-- begin:\n"
        + f"before: argument = {argument.value}\n"
        + "argument.value = argument.value * 2")

    argument.value *= 2

    print(f"after: argument.value = {argument.value}\n"
        + "-- end.")

value = 5
list = [2, 3, 6]
tuple = (1, 4, 8)
object = type('',(object,),{"value": 7})()

print(f"BEFORE: value = {value}")
function_receiving_value(value)
print(f"AFTER: value = {value}\n")

print(f"BEFORE: list[0] = {list[0]}")
function_receiving_list(list)
print(f"AFTER: list[0] = {list[0]}\n")

# print(f"BEFORE: tuple[0] = {tuple[0]}")
# function_receiving_tuple(tuple)
# print(f"AFTER: tuple[0] = {tuple[0]}\n")

print(f"BEFORE: object.value = {object.value}")
function_receiving_object(object)
print(f"AFTER: object.value = {object.value}\n")
